File: done/70_climbing_stairs.py
##!/usr/bin/env python
'''
    @author [mst]
    @file   70_climbing_stairs.py
    @brief  leetcode problems series
    You are climbing a staircase. It takes n steps to reach the top.
    Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?

    gains:
    -dp, memoization

    features, changelog:
    -2022.11: -submission

    @version 0.1 2022.11
'''


import time



# memoized dynamic programming solution: record every calculated number as we go
#
# sub: 95%T 57%S
class Solution:
    def climbStairs(self, n: int) -> int:

        memo = {1:1,2:2}

        for i in range(3,n+1):
            memo[i] = memo[i-1] + memo[i-2]

        return memo[n]

# straightforward recursive solution. this yielded time exeeding error to
# imply the need for memoization
#
# sub with lru_cache: 35%T 57%S
class Solution0:
    # from functools import lru_cache
    # @lru_cache(maxsize=5)
    def climbStairs(self, n: int) -> int:
        if n == 1 : return 1
        if n == 2 : return 2

        return self.climbStairs(n-1) + self.climbStairs(n-2)


################## DRIVER
def main():
    sol = Solution()

    # testing
    n = 38
    start = time.perf_counter() # [bp] time the start with performance counter
    res = sol.climbStairs(n)
    diff = time.perf_counter()-start
    print (f'task for {n=} resulted {res=} ran for {diff} secs')


if __name__ == ("__main__"):
    main()

File: done/70_climbing_stairs.cpp
/*
[mst] 205_isomorphic_strings.cpp
Given two strings s and t, determine if they are isomorphic.
Two strings s and t are isomorphic if the characters in s can be replaced to get t.
All occurrences of a character must be replaced with another character while
preserving the order of characters. No two characters may map to the same character, but a character may map to itself.

Gains:
-Hash map mapping

features, changelog:
-2022.12: -initial draft
*/


////////////////// LIBS
#include <iostream>            // usage of console prints
using namespace std;

#include <vector>
#include <unordered_map>

////////////////// DECL_IMPL


// memoized dynamic programming solution: record every calculated number as we go
// there was some issue with initializing:
//   vector<int> memo(n+1,0);
//
// sub: 100T 44S
class Solution {
public:
    int climbStairs(int n) {
		vector<int> memo;

		memo.push_back(0);	// member 0 is a dummy
		memo.push_back(1);
		memo.push_back(2);

        for (int i = 3; i<=n; i++)
			memo.push_back( memo[i-1] + memo[i-2]);

        return memo[n];
    }
};

////////////////// DRIVER
int main()
{
	Solution sol;
	cout << "result: " << sol.climbStairs(1) << endl;

	return 0;
}
